ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/UpnpNatManager.java

```diff
@@ -0,0 +1,487 @@
+          @Override
+          public void remoteDeviceAdded(final Registry registry, final RemoteDevice device) {
+            LOG.debug("UPnP Device discovered: " + device.getDetails().getFriendlyName());
+            inspectDeviceRecursive(device, recognizedServices.keySet());
+          }
+        };
+  }
+
+  /**
+   * Start the manager. Must not be in started state.
+   *
+   * @throws IllegalStateException if already started.
+   */
+  public synchronized void start() {
+    if (started) {
+      throw new IllegalStateException("Cannot start already-started service");
+    }
+
+    LOG.info("starting upnp service...");
+    upnpService.startup();
+    upnpService.getRegistry().addListener(registryListener);
+
+    // TODO: does jupnp do this automatically?
+    upnpService.getControlPoint().search(new STAllHeader());
+
+    initiateExternalIpQuery();
+
+    started = true;
+  }
+
+  /**
+   * Stop the manager. Must not be in stopped state.
+   *
+   * @throws IllegalStateException if stopped.
+   */
+  public synchronized void stop() {
+    if (!started) {
+      throw new IllegalStateException("Cannot stop already-stopped service");
+    }
+    upnpService.getRegistry().removeListener(registryListener);
+    upnpService.shutdown();
+
+    started = false;
+  }
+
+  /**
+   * Returns the first of the discovered services of the given type, if any.
+   *
+   * @param type is the type descriptor of the desired service
+   * @return the first instance of the given type, or null if none
+   */
+  private synchronized RemoteService getService(final String type) {
+    return recognizedServices.get(type);
+  }
+
+  /**
+   * Get the discovered WANIPConnection service, if any.
+   *
+   * @return the WANIPConnection Service if we have found it, or null.
+   */
+  public synchronized RemoteService getWANIPConnectionService() {
+    return getService(SERVICE_TYPE_WAN_IP_CONNECTION);
+  }
+
+  /**
+   * Get the local address on which we discovered our external IP address.
+   *
+   * <p>This can be useful to distinguish which network interface the external address was
+   * discovered on.
+   *
+   * @return the local address on which our GetExternalIP was discovered on, or null if no
+   *     GetExternalIP query has been performed successfully.
+   */
+  public String getDiscoveredOnLocalAddress() {
+    return this.discoveredOnLocalAddress;
+  }
+
+  /**
+   * Returns a CompletableFuture that will wait for the given service type to be discovered. No new
+   * query will be performed, and if the service has already been discovered, the future will
+   * complete in the very near future.
+   *
+   * @param serviceType is the service type to wait to be discovered.
+   * @return future that will return the desired service once it is discovered, or null if the
+   *     future is cancelled.
+   */
+  private CompletableFuture<RemoteService> discoverService(final String serviceType) {
+
+    return CompletableFuture.supplyAsync(
```

Comment:
I'm not sure I follow what you're saying, but I do understand that each call to `discoverService()` can potentially result in an infinite loop if (1) the `CompletableFuture` isn't explicitly stopped (e.g. `cancel()` is never called) and (2) no WANIPConnection service isn't found.

A simple solution here is to add a timeout internally to the `CompletableFuture`, perhaps as an optional argument to the function. Something like this:

``` java
    private CompletableFuture<RemoteService> discoverService(final String serviceType) {
+     return discoverService(serviceType, 180000);
+   }
+   private CompletableFuture<RemoteService> discoverService(final String serviceType, int timeoutMS) {

      return CompletableFuture.supplyAsync(
          () -> {

+           long start = System.currentTimeMillis();
+
            // wait until our thread is interrupted (assume future was cancelled)
            // or we discover the service
            while (!Thread.currentThread().isInterrupted()) {
              RemoteService service = getService(serviceType);
              if (null != service) {
                return service;
              } else {
+
+               long elapsed = System.currentTimeMillis() - start;
+               if (elapsed > timeoutMS) {
+                 throw new TimeoutException("Failed to observe WANIPConnection within "+ timeoutMS +" milliseconds");
+               }
+
                try {
                  Thread.sleep(100);
                } catch (InterruptedException e) {
                  // fall back through to "isInterrupted() check"
                }
              }
            }
            return null;
          });
    }
```

Reply:
The problem here is that this future is chained to other futures that are visible to the caller, so there isn't really any control over it. The timeout would prevent chewing up resources indefinitely, but it wouldn't help the caller avoid any side effects.

Reply:
I was thinking you could do something like: 
* Update `recognizedServices` to type `Map<String, CompletableFuture<RemoteService>>`
* Initialize key `SERVICE_TYPE_WAN_IP_CONNECTION` with an incomplete `CompletableFuture`
* Complete this future when the service is found in `inspectDeviceRecursive()` or log a warning if the future is already completed
* Update `discoverService` to return the future corresponding to the requested `serviceType`

Does that make sense?

Reply:
I see what you're getting after, and yes it makes sense. I think it solves two problems:

1) It prevents duplication of the `CompletableFuture`, so that only a single one will ever consume resources
2) It allows it to be tracked so that it can later be `cancel()`ed

What it doesn't solve is that it can effectively become an infinite loop (per my comment), but this is mitigated by (1) above.

Sounds good to me.

Reply:
Done: 0e08f04f119f2c14d48fa88e9ed7d90931d3a1a9